tests/test_decorators.py

```diff
@@ -23,47 +25,88 @@
 # Classes
 ##############################################################################
 
+class InvalidSetup(py_trees.behaviour.Behaviour):
+    def setup(self, timeout):
+        # A common mistake is to forget to return a boolean value
+        # Composite behaviours will at least check to make sure that
+        # their children do so and raise TypeError's if they fail
+        # to do so.
+        pass
 
-@py_trees.meta.failure_is_success
-class MustGoOnRegardless(py_trees.behaviours.Failure):
-    pass
+class DummyDecorator(py_trees.decorators.Decorator):
+    def __init__(self, child, name=py_trees.common.Name.AUTO_GENERATED):
+        super(DummyDecorator, self).__init__(name=name, child=child)
 
 ##############################################################################
 # Tests
 ##############################################################################
 
+def test_set_name():
+    console.banner("Set Name")
+    child = py_trees.behaviours.Success(name="Woohoo")
+    named_decorator = DummyDecorator(name="Foo", child=child)
+    no_named_decorator = DummyDecorator(child=child)
+    print("\n--------- Assertions ---------\n")
+    print("named_decorator.name == Foo")
+    assert(named_decorator.name == "Foo")
+    print("no_named_decorator.name == DummyDecorator\\n[Woohoo]")
+    assert(no_named_decorator.name == "DummyDecorator\n[Woohoo]")
+
+def test_invalid_child():
+    console.banner("Invalid Child")
+    print("\n--------- Assertions ---------\n")
+    print("TypeError is raised")
+    assert_raises(TypeError, DummyDecorator.__init__, child=5)
+
+def test_invalid_setup():
+    console.banner("Invalid Setup")
+    parent = py_trees.decorators.Decorator(
+        name="Decorator",
+        child=InvalidSetup(name="Invalid Setup")
+    )
+    print("\n--------- Assertions ---------\n")
+    print("TypeError is raised")
+    with assert_raises(TypeError) as context:
+        parent.setup(timeout=15)
+    print("TypeError has message with substring 'NoneType'")
+    assert("NoneType" in str(context.exception))
 
 def test_failure_is_success_tree():
     console.banner("Failure is Success Tree")
     root = py_trees.composites.Selector(name="Root")
     failure = py_trees.behaviours.Failure(name="Failure")
-    goon = MustGoOnRegardless(name="DontBeAfraidToBeTheGoon")
+    goon = py_trees.decorators.FailureIsSuccess(
```

Comment:
btw should we use a better name here like failure_is_success?

Reply:
Some of these are a relic of having fun with the tests when coding ... can't be serious all the time. These however are missing their context since all of the custom classes changed. Done.